text/0000-native-cli.md

```diff
@@ -0,0 +1,136 @@
+npm init neon-lib my-lib
+```
+
+Use whatever name you want for your library in place of `my-lib`.
+
+### Node apps
+
+To create a new Node app, use the command:
+
+```
+npm init neon-app my-app
+```
+
+Use whatever name you want for your library in place of `my-app`.
+
+### Electron apps
+
+To create a new Electron app, use the command:
+
+```
+npm init neon-electron-app my-app
+```
+
+Use whatever name you want for your library in place of `my-app`.
+
+## Building a Neon project
+
+All of the standard Neon project templates are set up to allow you to do a complete build from source with the `install` script, meaning that all you have to do to build your project is the standard `npm install` (or `npm i` for short) command from the project root.
+
+If you look at your `package.json`, you'll see that the install script delegates to Rust's `cargo` tool to build the native addon from Rust source. You can do this directly yourself by running `cargo build` (for a debug build) or `cargo build --release` (for a release build) from the project root.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Replacing `neon new`
+
+The reason we can now move past `neon new` is that `npm` has added support for custom templates via extensibility of the `npm init` command. We have pre-reserved npm packages called:
+
+- `create-neon-lib`
+- `create-neon-app`
+- `create-neon-electron-app`
+
+The extensible design of [`npm init`](https://docs.npmjs.com/cli/init) means that these packages are automatically used as project generators for the commands:
+
+- `npm init neon-lib <args...>`
+- `npm init neon-app <args...>`
+- `npm init neon-electron-app <args...>`
+
+respectively.
+
+## Replacing `neon build`
+
+There are a couple of key developments that will allow us to move past `neon build`:
+
+1. Once we move to the [N-API backend](https://github.com/neon-bindings/neon/issues/444), we no longer need to invalidate a Neon build when changing between versions of Node. This means we won't need to use the custom invalidation logic (which has proved to be fragile and a source of user frustration).
+2. We discovered that we can skip the final step of copying the built DLL into a `.node` file by replacing the JS boilerplate that loads the addons with a call to the lower-level but official [`process.dlopen()`](https://nodejs.org/api/process.html#process_process_dlopen_module_filename_flags) API.
+
+In order to make `cargo build` work from the root directory of a Neon project, we can put a `Cargo.toml` in the root directory that either defines a [Cargo workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html) or that shells out to running `cargo build` in the proper subdirectory. Alternatively, we might just try experimenting with flattening the directory structure so that the Cargo directory _is_ the root directory. We should experiment with prototypes to see what looks and feels best for the developer experience.
```

Comment:
My recommendation is to flatten the directory.

Placing a workspace `Cargo.toml` in the root adds an extra step for users with a monorepo that would like the neon module to be part of that workspace. First, they need to delete the `Cargo.toml` at the project root.

I'm not clearing how a build that shells out would work.

Reply:
Re: shelling out, I meant a `build.rs` in the root directory that shells out to `cargo build` from the subdirectory. But this is probably a really bad idea. :) I'll just delete the idea from the RFC.